features/glance.feature

```diff
@@ -0,0 +1,39 @@
+@glance
+Feature: Glance
+  As an administrator
+  I want to make sure that Glance component is working correctly
+
+  @create_kvm_image
+  Scenario: Creation of KVM image
+    Given KVM source image exists
+    And glance image does not exist
+    When I create new glance image based on jeos1
+    Then the image is listed as active
+    And this image has non-empty ID
+    And its ID can be used to show the image info
+    When I delete the image
```

Comment:
> Everything what is being done should be captured in the scenario text.

>> But that's not really true - there is a TON of setup done which is not captured by the scenario text (setup of the >> nodes, installation of SUSE Cloud, creation and application of Crowbar barclamp proposals...).  

Yes, setup and teardown of the environment in broader sense is not done inside the cucumber tests, here we need to synchronize with the deployment paths. This is really not captured by the scenarios and was not meant to. 

> I consider that as best practice, otherwise we end up with a lot of action hidden before the our eyes when reading the feature and spenario specifications.

>> As per above we already have a lot of stuff hidden.  But that's fine.  The scenario is only supposed to test one thing.

Cucumber tests are usually self-driven from inside and capture everything in step definitions. Unfortunately, with our approach we will always have two contexts - `mkcloud` infrastructure setup and "dumb" `cucumber` tests where the knowledge about the current setup is mandatory for the person who runs the tests. But this does not limit us for how to build the scenarios.


> I would not put constraints on what is actual testing and what not. If the scenario fails while the image is being deleted, it's straight forward to see it, more effort is actually not needed compared to a separate scenario having failed, the reason will be the same.

>> I disagree - the problem is that it breaks the test suite report, because if something breaks with image deletion, any scenario which has image deletion at the end will be reported as failing.  That's bad, because a) it makes the situation look worse than it actually is, and b) extra effort is required to check all failing scenarios and ascertain that they're all only due to a single issue with image deletion.

Maybe we both mean different things. I don't see the action of deletion as a generic event that is equal with all scenarios you need to check for. 

I went to [cucumber mailing list](https://groups.google.com/forum/#!forum/cukes) to get some information from those who designed this tool and found out that I was wrong with suggesting explicit image deletion at the end of a scenario. The recommended way here is to do the cleanup in the `Given` steps. It's important to add here, that we don't do any generic cleanup, we just need to get rid of a specific file if it exists. This is not (easily) doable in a separate scenario or hook due to scenario specific variables/config.

> Additionally, I would like to avoid using hooks as it could end up messy

>> Why?

Every scenario would expect new hook definition. As these hooks are defined in separate files in `features/support/some_file.rb` we would need to keep that always with sync with the scenarios and features, give these hooks some names, maybe create categories etc; but this is not the main issue here. The problem is with the specific config and variables for the scenarios - the hooks apply mostly for general things that are not dependent on any things defined within a scenario. Additionally hooks add more effort for newcomers to the testsuite to grasp it easily.

> and rather keep the scenarios' as simple as possible

>> Yes I agree - my proposal makes them simpler than they currently are, because it avoids polluting scenarios with extra steps for teardown.  For example look at http://benmabey.com/2008/05/19/imperative-vs-declarative-scenarios-in-user-stories.html - the declarative style is much more readable mainly because the scenario is much smaller.

As I mentioned above, if we cover such cases in the `Given` steps, we keep the scenarios clean.